Only in settings: 802-11-wireless.cpp~
diff -ur origsettings//802-11-wirelessdbus.h settings/802-11-wirelessdbus.h
--- origsettings//802-11-wirelessdbus.h	2009-01-22 17:50:04.000000000 +0100
+++ settings/802-11-wirelessdbus.h	2009-01-22 17:45:41.000000000 +0100
@@ -3,7 +3,7 @@
 #ifndef KNM_WIRELESSDBUS_H
 #define KNM_WIRELESSDBUS_H
 
-#include <nm-setting-802-11-wireless.h>
+#include <nm-setting-wireless.h>
 
 #include <kdebug.h>
 #include <kcoreconfigskeleton.h>
Only in settings: 802-11-wirelessdbus.h~
Only in settings: 802-11-wirelesspersistence.cpp~
diff -ur origsettings//802-11-wireless-securitydbus.h settings/802-11-wireless-securitydbus.h
--- origsettings//802-11-wireless-securitydbus.h	2009-01-22 17:50:04.000000000 +0100
+++ settings/802-11-wireless-securitydbus.h	2009-01-22 17:45:41.000000000 +0100
@@ -3,7 +3,7 @@
 #ifndef KNM_WIRELESSSECURITYDBUS_H
 #define KNM_WIRELESSSECURITYDBUS_H
 
-#include <nm-setting-802-11-wireless-security.h>
+#include <nm-setting-wireless-security.h>
 
 #include <kdebug.h>
 #include <kcoreconfigskeleton.h>
Only in settings: 802-11-wireless-securitydbus.h~
diff -ur origsettings//802-1xdbus.h settings/802-1xdbus.h
--- origsettings//802-1xdbus.h	2009-01-22 17:50:04.000000000 +0100
+++ settings/802-1xdbus.h	2009-01-22 17:45:41.000000000 +0100
@@ -3,7 +3,7 @@
 #ifndef KNM_SECURITY8021XDBUS_H
 #define KNM_SECURITY8021XDBUS_H
 
-#include <nm-setting-802-1x.h>
+#include <nm-setting-8021x.h>
 
 #include <kdebug.h>
 #include <kcoreconfigskeleton.h>
Only in settings: 802-1xdbus.h~
diff -ur origsettings//802-3-ethernet.cpp settings/802-3-ethernet.cpp
--- origsettings//802-3-ethernet.cpp	2009-01-22 17:50:04.000000000 +0100
+++ settings/802-3-ethernet.cpp	2009-01-22 17:48:41.000000000 +0100
@@ -5,7 +5,12 @@
 
 using namespace Knm;
 
-WiredSetting::WiredSetting() : Setting(Setting::Wired)
+WiredSetting::WiredSetting() : Setting(Setting::Wired),
+    mPort(WiredSetting::EnumPort::mii),
+    mDuplex(WiredSetting::EnumDuplex::full),
+    mSpeed(0),
+    mAutonegotiate(true),
+    mMtu(1500)
 {
 }
 
Only in settings: 802-3-ethernet.cpp~
diff -ur origsettings//802-3-ethernetdbus.h settings/802-3-ethernetdbus.h
--- origsettings//802-3-ethernetdbus.h	2009-01-22 17:50:04.000000000 +0100
+++ settings/802-3-ethernetdbus.h	2009-01-22 17:45:41.000000000 +0100
@@ -3,7 +3,7 @@
 #ifndef KNM_WIREDDBUS_H
 #define KNM_WIREDDBUS_H
 
-#include <nm-setting-802-3-ethernet.h>
+#include <nm-setting-wired.h>
 
 #include <kdebug.h>
 #include <kcoreconfigskeleton.h>
Only in settings: 802-3-ethernetdbus.h~
diff -ur origsettings//gsmpersistence.cpp settings/gsmpersistence.cpp
--- origsettings//gsmpersistence.cpp	2009-01-22 17:50:04.000000000 +0100
+++ settings/gsmpersistence.cpp	2009-01-22 17:45:41.000000000 +0100
@@ -18,7 +18,7 @@
 void GsmPersistence::load()
 {
   GsmSetting * setting = static_cast<GsmSetting *>(m_setting);
-  setting->setNumber(m_config->readEntry("number", QLatin1String( "*99#" )));
+  setting->setNumber(m_config->readEntry("number", "*99#"));
   setting->setUsername(m_config->readEntry("username", ""));
   // SECRET
   if (m_storageMode != ConnectionPersistence::Secure) {
Only in settings: gsmpersistence.cpp~
diff -ur origsettings//ipv4.cpp settings/ipv4.cpp
--- origsettings//ipv4.cpp	2009-01-22 17:50:04.000000000 +0100
+++ settings/ipv4.cpp	2009-01-22 17:56:18.000000000 +0100
@@ -5,7 +5,9 @@
 
 using namespace Knm;
 
-Ipv4Setting::Ipv4Setting() : Setting(Setting::Ipv4)
+Ipv4Setting::Ipv4Setting() : Setting(Setting::Ipv4),
+    mMethod(Ipv4Setting::EnumMethod::Automatic),
+    mIgnoredhcpdns(false)
 {
 }
 
Only in settings: ipv4.cpp~
Only in settings: .ipv4.cpp.swp
diff -ur origsettings//ipv4dbus.h settings/ipv4dbus.h
--- origsettings//ipv4dbus.h	2009-01-22 17:50:04.000000000 +0100
+++ settings/ipv4dbus.h	2009-01-22 17:45:41.000000000 +0100
@@ -3,7 +3,7 @@
 #ifndef KNM_IPV4DBUS_H
 #define KNM_IPV4DBUS_H
 
-#include <nm-setting-ipv4.h>
+#include <nm-setting-ip4-config.h>
 
 #include <kdebug.h>
 #include <kcoreconfigskeleton.h>
Only in settings: ipv4dbus.h~
Only in settings: .ipv4.h.swp
diff -ur origsettings//serialpersistence.cpp settings/serialpersistence.cpp
--- origsettings//serialpersistence.cpp	2009-01-22 17:50:04.000000000 +0100
+++ settings/serialpersistence.cpp	2009-01-22 17:45:41.000000000 +0100
@@ -20,7 +20,7 @@
   SerialSetting * setting = static_cast<SerialSetting *>(m_setting);
   setting->setBaud(m_config->readEntry("baud", 0));
   setting->setBits(m_config->readEntry("bits", 8));
-  setting->setParity(m_config->readEntry("parity", QLatin1String( "None" )));
+  setting->setParity(m_config->readEntry("parity", "None"));
   setting->setStopbits(m_config->readEntry("stopbits", 1));
   setting->setSenddelay(m_config->readEntry("senddelay", 0));
 }
Only in settings: serialpersistence.cpp~
Only in settings: .svn
